src/steps/converters.ts

```
@@ -41,7 +53,7 @@ export function createUserEntity(user: any): Entity {
         ...convertCommonServiceNowProperties(user),
         _class: Entities.USER._class,
         _type: Entities.USER._type,
-        _key: user.sys_id,
+        _key: getUserKey(user.sys_id, user.user_name),
```

Do you have examples of where this fixed the duplicate key issue? I'm surprised the `id` wasn't unique.

---

Hey @zemberdotnet, we aren't 100% sure if this will fix the issue but we thought it was worth a try - weren't aware changing `_key` has negative impact as a whole on the other hand.

The code surrounding user ingestion is very simple:

Client method part:
```
return this.iterateTableResources({
      table: ServiceNowTable.USER,
      callback,
});
```

Step function part:
```
await client.iterateUsers(async (user) => {
    const userEntity = await jobState.addEntity(createUserEntity(user));

    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.HAS,
        from: accountEntity,
        to: userEntity,
      }),
    );
});
```

Some ideas that came to mind were either to increase the `_key` complexity or to maybe skip (truly) duplicate users but it's not like we're expecting them by default and that alone might hide some other bug that's causing this.

Do you have some ideas what you might try doing here?

---

Hey @eXtremeX 👋 Thank you for the detailed explanation.
> weren't aware changing _key has negative impact as a whole on the other hand.
In this case, it's probably okay. It will cause all entities to be fully deleted and recreated, but this probably doesn't produce enough data for that to be a huge concern.

I'm good to try this approach if we can address the other comment I have below.